crates/loro-internal/src/encoding/arena.rs

```diff
@@ -0,0 +1,463 @@
+use std::{borrow::Cow, ops::Deref};
+
+use crate::InternalString;
+use fxhash::FxHashSet;
+use itertools::Itertools;
+use loro_common::{ContainerID, ContainerType, Counter, LoroError, LoroResult, PeerID};
+use serde::{Deserialize, Serialize};
+use serde_columnar::{columnar, ColumnarError};
+
+use super::encode_reordered::{PeerIdx, ValueRegister, MAX_DECODED_SIZE};
+
+pub(super) fn encode_arena(
+    peer_ids_arena: Vec<u64>,
+    containers: ContainerArena,
+    keys: Vec<InternalString>,
+    deps: DepsArena,
+    position_arena: PositionArena,
+    tree_id_arena: TreeIDArena,
+    state_blob_arena: &[u8],
+) -> Vec<u8> {
+    let peer_ids = PeerIdArena {
+        peer_ids: peer_ids_arena,
+    };
+
+    let key_arena = KeyArena { keys };
+    let encoded = EncodedArenas {
+        peer_id_arena: &peer_ids.encode(),
+        container_arena: &containers.encode(),
+        key_arena: &key_arena.encode(),
+        deps_arena: &deps.encode(),
+        position_arena: &position_arena.encode(),
+        tree_id_arena: &tree_id_arena.encode(),
+        state_blob_arena,
+    };
+
+    encoded.encode_arenas()
+}
+
+pub struct EncodedRegisters<'a> {
+    pub(super) peer: ValueRegister<PeerID>,
+    pub(super) key: ValueRegister<InternalString>,
+    pub(super) container: ValueRegister<ContainerID>,
+    pub(super) tree_id: ValueRegister<EncodedTreeID>,
+    pub(super) position: either::Either<FxHashSet<&'a [u8]>, ValueRegister<&'a [u8]>>,
+}
+
+impl<'a> EncodedRegisters<'a> {
+    pub(crate) fn convert_position(&mut self) {
```

Comment:
I think it's better to call `position` as `fractional_index` here directly to reduce the confusion.

Reply:
agree